packages/openapi-v3/src/parameter-decorator.ts

```diff
@@ -46,263 +27,498 @@ const paramDecoratorStyle = Symbol('ParamDecoratorStyle');
+      // The design-time type is `Object` for `any`
+      if (paramType != null && paramType !== Object && paramType !== Array) {
         throw new Error(
-          'Mixed usage of @param at method/parameter level' +
-            ' is not allowed.',
+          `The parameter type is set to 'array' but the JavaScript type is ${
+            paramType.name
+          }`,
         );
       }
-      targetWithParamStyle[paramDecoratorStyle] = 'method';
-      RestMethodParameterDecoratorFactory.createDecorator<ParameterObject>(
-        OAI2Keys.METHODS_WITH_PARAMETERS_KEY,
-        paramSpec,
-      )(target, member, descriptorOrIndex);
     }
+
+    ParameterDecoratorFactory.createDecorator<ParameterObject>(
+      OAI3.PARAMETERS_KEY,
+      paramSpec,
+    )(target, member, index);
   };
 }
 
-class RestMethodParameterDecoratorFactory extends MethodParameterDecoratorFactory<
-  ParameterObject
-> {}
+/**
+ * The `type` and `format` inferred by a common name of OpenAPI 3.0.0 data type
+ * reference link:
+ * https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md#data-types
+ */
+const typeAndFormatMap = {
+  integer: {type: 'integer', format: 'int32'},
+  long: {type: 'integer', format: 'int64'},
+  float: {type: 'number', format: 'float'},
+  double: {type: 'number', format: 'double'},
+  byte: {type: 'string', format: 'byte'},
+  binary: {type: 'string', format: 'binary'},
+  date: {type: 'string', format: 'date'},
+  dateTime: {type: 'string', format: 'date-time'},
+  password: {type: 'string', format: 'password'},
+};
 
+/**
+ * Shortcut parameter decorators
+ */
 export namespace param {
   export const query = {
     /**
-     * Define a parameter of "string" type that's read from the query string.
-     *
+     * Define a parameter of "integer" type that's read from the query string.
+     * Usage: @param.query.string('paramName')
+     * 
      * @param name Parameter name.
      */
-    string: createParamShortcut('query', 'string'),
-
+    string: createParamShortCut('query', 'string'),
     /**
      * Define a parameter of "number" type that's read from the query string.
-     *
+     * Usage: @param.query.number('paramName')
+     * 
      * @param name Parameter name.
      */
-    number: createParamShortcut('query', 'number'),
-
+    number: createParamShortCut('query', 'number'),
+    /**
+     * Define a parameter of "boolean" type that's read from the query string.
+     * Usage: @param.query.boolean('paramName')
+     * 
+     * @param name Parameter name.
+     */
+    boolean: createParamShortCut('query', 'boolean'),
     /**
      * Define a parameter of "integer" type that's read from the query string.
-     *
+     * Usage: @param.query.integer('paramName')
+     * 
      * @param name Parameter name.
      */
-    integer: createParamShortcut('query', 'integer'),
-
+    integer: createParamShortCut(
+      'query',
+      typeAndFormatMap.integer.type,
+      typeAndFormatMap.integer.format,
```

Comment:
I am proposing to rework `createParamShortCut`  to accept and object with type & format as the second argument.

```ts
// flavor 1 - type specified as string
boolean: createParamShortCut('query', 'boolean')

// flavour 2 - type specified as an object
integer: createParamShortCut('query', typeAndFormatMap.integer),
```

Also please keep the helper function name with lowercase 'c', i.e. `createParamShortcut`, it will make the diff size smaller among other things.

Reply:
Good point! refactored it in commit https://github.com/strongloop/loopback-next/pull/1046/commits/0acc43cf0ecb3137d78a776689908ceb56cac3aa